### PR TITLE
Expire IMap key before locking

### LINKED ISSUE #13272: IMap.put(String,Object,long,TimeUnit) broken starting with 3.9.2
After ttl exceeded the  map entry is still there with all hazelcast versions since 3.9.2. 
3.9.1 and earlier are ok.

### PR DESCRIPTION
**Pull request by vbekiaris**
Introduces SPI for services that need to
execute some code before locking
a key.

Fixes #13272

### CODE REVIEW

File: hazelcast/src/main/java/com/hazelcast/concurrent/lock/operations/LockOperation.java [hunk 1]
```diff
@@ -45,6 +46,12 @@ public LockOperation(ObjectNamespace namespace, Data key, long threadId, long le
 
     @Override
     public void run() throws Exception {
+        // if service is a LockInterceptorService, notify it a key is about to be locked
+        Object targetService = getNodeEngine().getService(namespace.getServiceName());
+        if (targetService instanceof LockInterceptorService) {
+            ((LockInterceptorService) targetService).onBeforeLock(namespace.getObjectName(), key);
```
Comment (ahmetmircik) [inline hazelcast/src/main/java/com/hazelcast/concurrent/lock/operations/LockOperation.java]: This code-block can also be put `LockBackupOperation`?

File: hazelcast/src/main/java/com/hazelcast/spi/LockInterceptorService.java [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+import com.hazelcast.nio.serialization.Data;
+
+/**
+ * Interface to be implemented by services that need to intercept lock operation
+ * for distributed objects it manages.
+ */
+public interface LockInterceptorService {
+
+    void onBeforeLock(String distributedObjectName, Data key);
```
Comment (ahmetmircik) [inline hazelcast/src/main/java/com/hazelcast/spi/LockInterceptorService.java]: Maybe passing Object type can make contract more open to other use cases:
`void onBeforeLock(String distributedObjectName, Object param)`?

File: hazelcast/src/test/java/com/hazelcast/spi/LockInterceptorServiceTest.java [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+
+import java.util.Properties;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import static com.hazelcast.util.ConcurrencyUtil.getOrPutIfAbsent;
+import static org.junit.Assert.assertEquals;
+
+public class LockInterceptorServiceTest extends HazelcastTestSupport {
```
Comment (ahmetmircik) [inline hazelcast/src/test/java/com/hazelcast/spi/LockInterceptorServiceTest.java]: category? test-runner?

**Comment by ahmetmircik** [comment]
can we put some test for the real issue?

File: hazelcast/src/main/java/com/hazelcast/concurrent/lock/operations/LockOperation.java (hunk 1 above)
Comment (vbekiaris) [reply hazelcast/src/main/java/com/hazelcast/concurrent/lock/operations/LockOperation.java]: +1, pulled as separate method to `AbstractLockOperation` and used in both `LockOperation` and `LockBackupOperation`.

File: hazelcast/src/main/java/com/hazelcast/spi/LockInterceptorService.java (hunk 2 above)
Comment (vbekiaris) [reply hazelcast/src/main/java/com/hazelcast/spi/LockInterceptorService.java]: Converted type of key to type parameter

File: hazelcast/src/test/java/com/hazelcast/spi/LockInterceptorServiceTest.java (hunk 3 above)
Comment (vbekiaris) [reply hazelcast/src/test/java/com/hazelcast/spi/LockInterceptorServiceTest.java]: good catch

**Comment by vbekiaris** [comment]
@ahmetmircik the test for the original `IMap.lock` issue has been contributed in a separate PR (#13278). I expect to have the test PR merged and then un-ignored in this PR.

**Pull request merged**